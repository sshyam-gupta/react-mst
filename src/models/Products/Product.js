import { types } from 'mobx-state-tree';

const Product = types
  .model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, '')
  })
  .actions(self => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice;
    },
    changeImage(newImage) {
      self.image = newImage;
    }
  }));

export default Product;
