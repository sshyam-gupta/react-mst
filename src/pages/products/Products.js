import React, { Component } from 'react';

import { Container } from './elements';
import { inject } from 'mobx-react';
import ProductList from './ProductList';
import AddProduct from './AddProduct';

class Products extends Component {
  render() {
    const { products } = this.props.productStore;

    return (
      <Container>
        <AddProduct />
        <ProductList products={products} />
      </Container>
    );
  }
}

export default inject('productStore')(Products);
