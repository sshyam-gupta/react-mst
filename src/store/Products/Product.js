import { types, getParent } from 'mobx-state-tree';

const Product = types
  .model({
    name: types.string,
    price: types.number,
    image: types.optional(types.string, '')
  })
  .actions(self => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice;
    },
    changeImage(newImage) {
      self.image = newImage;
    },
    removeProduct() {
      getParent(self, 2).remove(self);
    }
  }));

export default Product;
