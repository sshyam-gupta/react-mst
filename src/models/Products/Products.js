import { types } from 'mobx-state-tree';

import Product from './Product';

const Products = types
  .model({
    products: types.optional(types.array(Product), [])
  })
  .actions(self => ({
    add(product) {
      self.products.push(product);
    }
  }))
  .views(self => ({
    get totalPrice() {
      return self.products.reduce((sum, product) => sum + product.price, 0);
    }
  }));

export default Products;
