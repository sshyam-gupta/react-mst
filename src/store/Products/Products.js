import { types, destroy } from 'mobx-state-tree';

import Product from './Product';

const Products = types
  .model({
    products: types.optional(types.array(Product), [])
  })
  .actions(self => ({
    add(product) {
      self.products.push(product);
    },
    remove(product) {
      destroy(product);
    }
  }))
  .views(self => ({
    get totalPrice() {
      return self.products.reduce((sum, product) => sum + product.price, 0);
    }
  }));

export default Products;
