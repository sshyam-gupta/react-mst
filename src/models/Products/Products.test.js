import Product from './Product';
import Products from './Products';
import { getSnapshot, onSnapshot } from 'mobx-state-tree';

it('Can add new product', () => {
  let item = Products.create();

  let states = [];

  onSnapshot(item, snapshot => {
    states.push(snapshot);
  });

  item.add(
    Product.create({
      name: 'test',
      price: 10,
      image: 'sdds'
    })
  );

  expect(item.products.length).toBe(1);

  expect(item.products[0].name).toBe('test');

  item.products[0].changeName('Shyam');
  expect(item.products[0].name).toBe('Shyam');

  expect(getSnapshot(item)).toEqual({
    products: [
      {
        name: 'Shyam',
        price: 10,
        image: 'sdds'
      }
    ]
  });

  expect(getSnapshot(item)).toMatchSnapshot();
  expect(states).toMatchSnapshot();
});

it('gives total price', () => {
  let item = Products.create({
    products: [
      {
        name: 'test',
        price: 10
      },
      {
        name: 'test1',
        price: 100
      }
    ]
  });
  expect(item.totalPrice).toBe(110);

  item.products[0].changePrice(20);

  expect(item.totalPrice).toBe(120);
});
