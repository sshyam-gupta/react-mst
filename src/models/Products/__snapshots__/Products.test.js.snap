// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can add new product 1`] = `
Object {
  "products": Array [
    Object {
      "image": "sdds",
      "name": "Shyam",
      "price": 10,
    },
  ],
}
`;

exports[`Can add new product 2`] = `
Array [
  Object {
    "products": Array [
      Object {
        "image": "sdds",
        "name": "test",
        "price": 10,
      },
    ],
  },
  Object {
    "products": Array [
      Object {
        "image": "sdds",
        "name": "Shyam",
        "price": 10,
      },
    ],
  },
]
`;
